### ISSUE TITLE
[onert/train] support AveragePool2D for training

### ISSUE THREAD

**Issue by jyoungyun** [Samsung/ONE#13829]
## What

Let's support AveragePool2D for training.

## Why?

The InceptionV3 model from TensorFlow applications has AveragePool2D operation. However, onert does not support training this operation. PoolLayer is already implemented and only supports `MaxPool2D`. This issue can expand the pool type to support `AveragePool2D`.

**Comment by icodo98** [comment]
I made a draft for this issue. https://github.com/Samsung/ONE/pull/14059
PTAL 😊

**Comment by icodo98** [comment]
Here is sample model and result.
[averagePool.zip](https://github.com/user-attachments/files/17106330/averagePool.zip)

![image](https://github.com/user-attachments/assets/faa263bb-1977-4c73-a691-315b9b06dfd7)

**Comment by icodo98** [comment]
from https://github.com/Samsung/ONE/pull/14086#discussion_r1776162344

while replacing code to use existing forwarding function, I've got a question.

for `MaxPool2D` in training, it's able to forward output from padding area.

https://github.com/Samsung/ONE/blob/06a536b634f08145f1656adc7664b8099776f9a6/compute/cker/src/train/MaxPool.test.cc#L205-L211

for average pool, I think the output will be

```
    /**
     * input_with_padding:             expected_output:
     *
     *    4   8                              0    0    0
     *    9   2            -(forward)->      0  5.75   0
     *                                       0    0    0
     */
```

But with current forwarding `AveragePool` function,  above test case fails.

https://github.com/Samsung/ONE/blob/06a536b634f08145f1656adc7664b8099776f9a6/compute/cker/include/cker/operation/AveragePool.h#L94-L96

`assert(out_count >0)` prevent output induced from padded layer. 

So, is our policy to not induce output from padded area? or is it bug?

/cc @ragmani @zetwhite

**Comment by ragmani** [comment]
> So, is our policy to not induce output from padded area? or is it bug?

It's the former. There is no case that circle model has paddings like the above test. And onert only allows padding type as "valid" and "same" for Pool ops now. The two padding types never induce Pool ops to have paddings like the above test.

**Comment by zetwhite** [comment]
> And onert only allows padding type as "valid" and "same" for Pool ops now 

Ah, I've had a different understanding. 

I thought that "explicit" type is also possible for onert Pool ops. Because while generating kernel for pool operator, KernelGenerator allows explicit pool params. 

* KernelGenerator.cc, calculate padding 
https://github.com/Samsung/ONE/blob/a1e988429d416ac9b271c7347f0efb4fc09f3dad/runtime/onert/backend/train/KernelGenerator.cc#L483-L484 

* `calculatePadding()` it allows explicit padding type 
https://github.com/Samsung/ONE/blob/a1e988429d416ac9b271c7347f0efb4fc09f3dad/runtime/onert/core/src/ir/Padding.cc#L137-L157


------------------------------------------------

>  There is no case that circle model has paddings like the above test. 

Maybe Does it mean that the frontend does not allow explicit type?

**Comment by ragmani** [comment]
> Maybe Does it mean that the frontend does not allow explicit type?

I guess that training also does not allow explicit type for pool ops. As far as I know, explicit type may be allowed only in the case of pad op now.

**Comment by icodo98** [comment]
> As far as I know, explicit type may be allowed only in the case of pad op now.

Then, I'll remove that test case from PR.

**Comment by icodo98** [comment]
Even after #14149 , InceptionV3 model is not supported for training because `Concat` operation is not trainable in One yet.
Can I open new issue for this?

**Comment by shs-park** [comment]
Sure, just make a link to this issue as related =)

**Comment by ranykamel** [comment]
Hello! I'm interested in taking on this issue if it's still relevant and available. This is my first time contributing to this repository, so I'd appreciate any advice or pointers to help me get started.
Thanks!

**Comment by llFreetimell** [comment]
@ranykamel 

Thanks for your interests on this repository.
Unfortunately, we decided to stop merging any PRs (https://github.com/Samsung/ONE/issues/16479), which is no use assigning this issue to you or creating a draft PR.
I'm sorry I don't have better news.

**Comment by ranykamel** [comment]
> @ranykamel 
> 
> Thanks for your interests on this repository.
> Unfortunately, we decided to stop merging any PRs (https://github.com/Samsung/ONE/issues/16479), which is no use assigning this issue to you or creating a draft PR.
> I'm sorry I don't have better news.

Thank you for letting me know @llFreetimell. The timing is quite unfortunate. Hopefully I will get to work on this repository sometime in the future.
